### DISCUSSION TITLE
Add BufferWithCount or an overload to Buffer()

### DISCUSSION THREAD

**Discussion by softlion** [dotnet/reactive#1692] [Ideas]
I know this "idea" has already discussions on rx, but i feel those discussions were aside from the real need. So i'm opening a new one.

BufferWithCount(N) should be a rolling (ie: overlapping) buffer emitting the N last (buffered) items each time one new item is added.  
Currently, Buffer(N) is only a non overlapping buffer, emitting distinct group of items.

The current behavior of Buffer makes difficult to compare 2 consecutive values.  And to emit a new item only when a specific change occurs between those consecutive values.

The idea is to simplify this code:

```c#
var queue = new List<AdapterStatus>();
this.WhenAnyValue(x => x.ViewModel.AdapterStatus)
    .StartWith(AdapterStatus.Unknown)
    .Distinct()
    .Do(value =>
    {
        queue.Add(value);
        if(queue.Count > 2)
            queue.RemoveAt(0);
    })
    .Where(s => queue.Count>=2 && queue[^1] == AdapterStatus.Available && queue[^2] != AdapterStatus.Available) 
```

Into this:

```c#
this.WhenAnyValue(x => x.ViewModel.AdapterStatus)
    .StartWith(AdapterStatus.Unknown)
    .Distinct()
    .BufferWithCount(2) //Keep this curious name to be 'standard' with other languages
    .Where(s => queue[^1] == AdapterStatus.Available && queue[^2] != AdapterStatus.Available) 
```

Does this make sense ?

**Comment by akarnokd** [comment]
I'm pretty sure there is a standard `Buffer` overload with count and skip where you get this overlapping pattern:

```cs
Observable.Range(1, 10)
            .Buffer(2, 1)
            .Subscribe(v => Console.WriteLine(string.Join(",", v)));
```

prints

```
1,2
2,3
3,4
4,5
5,6
6,7
7,8
8,9
9,10
10
```

**Reply by softlion** [reply]
You are right. The names of nearly all methods and parameters should be reviewed at some time in all RX. They are all so misleading.

**Comment by dwcullop** [comment]
> The current behavior of Buffer makes difficult to compare 2 consecutive values. And to emit a new item only when a specific change occurs between those consecutive values.

If that is all you need, there is already the `DistinctUntilChanged` operator that will suppress events if the value is the same as the last one.  There is even an overload that will take `IEqualityComparer<T>` so you can control what "same" means.
